RecoTauTag/RecoTau/plugins/DPFIsolation.cc

```diff
@@ -0,0 +1,387 @@
+            float lepRecoPt = tau.pt();
+            float lepRecoPz = std::abs(tau.pz());
+
+            // Use of setZero results in warnings in eigen library during compilation.
+            //tensor.flat<float>().setZero();
+            const unsigned n_inputs = getNumberOfParticles(graphVersion) * GetNumberOfFeatures(graphVersion);
+            for(unsigned input_idx = 0; input_idx < n_inputs; ++input_idx)
+                tensor.flat<float>()(input_idx) = 0;
+
+            unsigned int iPF = 0;
+            const unsigned max_iPF = getNumberOfParticles(graphVersion);
+
+            std::vector<unsigned int> sorted_inds(pfcands->size());
+            std::size_t n = 0;
+            std::generate(std::begin(sorted_inds), std::end(sorted_inds), [&]{ return n++; });
+
+            std::sort(std::begin(sorted_inds), std::end(sorted_inds),
+            [&](int i1, int i2) { return pfcands->at(i1).pt() > pfcands->at(i2).pt(); } );
+
+            for(size_t pf_index = 0; pf_index < pfcands->size() && iPF < max_iPF; pf_index++) {
+                pat::PackedCandidate p = pfcands->at(sorted_inds.at(pf_index));
+                float deltaR_tau_p =  deltaR(p.p4(),tau.p4());
+
+                if (p.pt() < 0.5) continue;
+                if (p.fromPV() < 0) continue;
+                if (deltaR_tau_p > 0.5) continue;
+                if (p.fromPV() < 1 && p.charge() != 0) continue;
+                pfCandPt = p.pt();
+                pfCandPtRel = p.pt()/lepRecoPt;
+
+                pfCandDr = deltaR_tau_p;
+                pfCandDEta = std::abs(tau.eta() - p.eta());
+                pfCandDPhi = std::abs(deltaPhi(tau.phi(), p.phi()));
+                pfCandEta = p.eta();
+                pfCandIsBarrel = (std::abs(pfCandEta) < 1.4);
+                pfCandPz = std::abs(std::sinh(pfCandEta)*pfCandPt);
+                pfCandPzRel = std::abs(std::sinh(pfCandEta)*pfCandPt)/lepRecoPz;
+                pfCandPdgID = std::abs(p.pdgId());
+                pfCandCharge = p.charge();
+                pfCandDVx_1 = p.vx() - pvx;
+                pfCandDVy_1 = p.vy() - pvy;
+                pfCandDVz_1 = p.vz() - pvz;
+
+                pfCandD_1 = std::sqrt(pfCandDVx_1*pfCandDVx_1 + pfCandDVy_1*pfCandDVy_1 + pfCandDVz_1*pfCandDVz_1);
+
+                if (pfCandCharge != 0 and p.hasTrackDetails()){
+                    pfCandDz      = p.dz();
+                    pfCandDzErr   = p.dzError();
+                    pfCandDzSig   = (std::abs(p.dz()) + 0.000001)/(p.dzError() + 0.00001);
+                    pfCandD0      = p.dxy();
+                    pfCandD0Err   = p.dxyError();
+                    pfCandD0Sig   = (std::abs(p.dxy()) + 0.000001)/ (p.dxyError() + 0.00001);
+                    pfCandPixHits = p.numberOfPixelHits();
+                    pfCandHits    = p.numberOfHits();
+                    pfCandLostInnerHits = p.lostInnerHits();
+                } else {
+                    float disp = 1;
+                    int psudorand = p.pt()*1000000;
```

Comment:
@ocolegrove please comment

Reply:
In my understanding it is that DNN prefers that some values are always set for each variable even if the values have not any meaning for a given category, here track parameters for particles w/o tracks. In this particular case, the values are set to values on the boundaries of allowed ranges, i.e. either at min or max value. Some variables (here IPs) are signed thus sign is added at random basis to not prefer negative or positive boundary. After said this, at least naivety IP's values set to zero should also work, however it could require retraining. @ocolegrove, am I right?